### PR TITLE
Fixed #34112 -- Added async-compatible interface to Model methods.

### PR DESCRIPTION
**Pull request by DevilsAutumn**
[Ticket #34112](https://code.djangoproject.com/ticket/34112)

### CODE REVIEW

File: django/db/models/base.py [hunk 1]
```diff
@@ -1111,6 +1126,12 @@ def delete(self, using=None, keep_parents=False):
 
     delete.alters_data = True
 
+    async def adelete(self, using=None, keep_parents=False):
+        return await sync_to_async(self.delete)(
+            using=using,
+            keep_parents=keep_parents,
+        )
```
Comment (felixxm) [inline django/db/models/base.py:1133]: As far as I'm aware, `adelete()` and `asave()` should have `alters_data` set to `True`.

File: docs/topics/async.txt [hunk 2]
```diff
@@ -91,10 +91,21 @@ Detailed notes can be found in :ref:`async-queries`, but in short:
 * ``async for`` is supported on all QuerySets (including the output of
   ``values()`` and ``values_list()``.)
 
+Django also supports some asynchronous model methods that use the database::
+
+    async def make_book(...):
+        book = Book(...)
+        await book.asave(using="secondary")
+
+
```
Comment (felixxm) [inline docs/topics/async.txt:100]: Chop blank line
```suggestion
```

File: tests/async/test_async_queryset.py [hunk 3]
```diff
@@ -243,6 +243,6 @@ async def test_aexplain(self):
                         self.fail(f"QuerySet.aexplain() result is not valid JSON: {e}")
 
     async def test_raw(self):
-        sql = "SELECT id, field FROM async_queryset_simplemodel WHERE created=%s"
```
Comment (felixxm) [inline tests/async/test_async_queryset.py:246]: Removing `async_queryset` app is unrelated cleanup. Please move it to a separate commit/PR.

File: django/db/models/base.py (hunk 1 above)
Comment (DevilsAutumn) [reply django/db/models/base.py:1133]: ``alters_data`` is already True for ``delete`` and  ``save``, do i need to do it again for ``asave`` and ``adelete``?

Comment (felixxm) [reply django/db/models/base.py:1133]: > `alters_data` is already True for `delete` and `save`, do i need to do it again for `asave` and `adelete`?

Yes, because the template system is not aware what methods are called inside.

**Comment by adamchainz** [comment]
(Carrying on from #16235.)

File: tests/async/models.py [hunk 4]
```diff
@@ -1,5 +1,9 @@
 from django.db import models
 
 
+class RelatedModel(models.Model):
+    simple = models.ForeignKey("SimpleModel", models.CASCADE, null=True)
+
+
```
Comment (adamchainz) [inline tests/async/models.py:7]: this model doesn't need to be in this commit now

```suggestion
```

**Review by adamchainz — COMMENTED** [review]
LGTM asides from the below

File: tests/async/models.py (hunk 4 above)
Comment (DevilsAutumn) [reply tests/async/models.py]: Thanks, missed it somehow. 😢

**Comment by DevilsAutumn** [comment]
Thanks @adamchainz . For async_queryset module cleanup, should i create a new PR or add in a new commit in this PR only?

**Comment by adamchainz** [comment]
Either would be acceptable. Squash your existing commits into one titled `Fixed #34112 -- Added async interface to model methods.` (including final full stop). Then you can add another PR here titled `Refs #34112 -- Moved async queryset tests into tests/async.`.

**Comment by DevilsAutumn** [comment]
After squashing do i have to push it on github?

**Comment by adamchainz** [comment]
If you’ve squashed properly, there won’t be merge conflicts, but you will need to force push. Here’s a tutorial I wrote on squashing branches: https://adamj.eu/tech/2022/03/25/how-to-squash-and-rebase-a-git-branch/

**Comment by DevilsAutumn** [comment]
> If you’ve squashed properly, there won’t be merge conflicts, but you will need to force push. Here’s a tutorial I wrote on squashing branches: https://adamj.eu/tech/2022/03/25/how-to-squash-and-rebase-a-git-branch/

Thankyou so much @adamchainz , learned a lot from you. Hope everything is fine this time. :)

**Comment by felixxm** [comment]
@DevilsAutumn Thanks :+1:

**Comment by adamchainz** [comment]
Great work, thanks!
